File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/integrations/PlaceholderIntegrationPAPI.java
package com.willfp.eco.internal.integrations;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.integrations.placeholder.PlaceholderIntegration;
import com.willfp.eco.core.integrations.placeholder.PlaceholderManager;
import me.clip.placeholderapi.PlaceholderAPI;
import me.clip.placeholderapi.expansion.PlaceholderExpansion;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class PlaceholderIntegrationPAPI extends PlaceholderExpansion implements PlaceholderIntegration {
    /**
     * The linked plugin.
     */
    private final EcoPlugin plugin;

    /**
     * Create a new PlaceholderAPI integration.
     *
     * @param plugin The plugin to manage placeholders for.
     */
    public PlaceholderIntegrationPAPI(@NotNull final EcoPlugin plugin) {
        this.plugin = plugin;
    }

    @Override
    public boolean persist() {
        return true;
    }

    @Override
    public boolean canRegister() {
        return true;
    }

    @Override
    public @NotNull String getAuthor() {
        return String.join(", ", plugin.getDescription().getAuthors());
    }

    @Override
    public @NotNull String getIdentifier() {
        return plugin.getDescription().getName().toLowerCase();
    }

    @Override
    public @NotNull String getVersion() {
        return plugin.getDescription().getVersion();
    }

    @Override
    public String onPlaceholderRequest(@Nullable final Player player,
                                       @NotNull final String identifier) {
        return PlaceholderManager.getResult(player, identifier);
    }

    @Override
    public void registerIntegration() {
        this.register();
    }

    @Override
    public String getPluginName() {
        return "PlaceholderAPI";
    }

    @Override
    public String translate(@NotNull final String text,
                            @Nullable final Player player) {
        return PlaceholderAPI.setPlaceholders(player, text);
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/integrations/PlaceholderIntegrationPAPI.kt
package com.willfp.eco.internal.integrations

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.integrations.placeholder.PlaceholderIntegration
import com.willfp.eco.core.integrations.placeholder.PlaceholderManager
import me.clip.placeholderapi.PlaceholderAPI
import me.clip.placeholderapi.expansion.PlaceholderExpansion
import org.bukkit.entity.Player

class PlaceholderIntegrationPAPI(private val plugin: EcoPlugin) : PlaceholderExpansion(), PlaceholderIntegration {
    override fun persist(): Boolean {
        return true
    }

    override fun canRegister(): Boolean {
        return true
    }

    override fun getAuthor(): String {
        return java.lang.String.join(", ", plugin.description.authors)
    }

    override fun getIdentifier(): String {
        return plugin.description.name.lowercase()
    }

    override fun getVersion(): String {
        return plugin.description.version
    }

    override fun onPlaceholderRequest(
        player: Player?,
        identifier: String
    ): String {
        return PlaceholderManager.getResult(player, identifier, plugin)
    }

    override fun registerIntegration() {
        register()
    }

    override fun getPluginName(): String {
        return "PlaceholderAPI"
    }

    override fun translate(
        text: String,
        player: Player?
    ): String {
        return PlaceholderAPI.setPlaceholders(player, text)
    }

    override fun findPlaceholdersIn(text: String): MutableList<String> {
        val placeholders = mutableListOf<String>()
        val matcher = PlaceholderAPI.getPlaceholderPattern().matcher(text)
        while (matcher.find()) {
            placeholders.add(matcher.group())
        }

        return placeholders
    }
}